src/coreclr/nativeaot/System.Private.CoreLib/src/System/CrashInfo.cs

```diff
@@ -0,0 +1,389 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Internal.DeveloperExperience;
+
+namespace System
+{
+    internal unsafe class CrashInfo
+    {
+        /// <summary>
+        /// The kind or reason of crash for the triage JSON
+        /// </summary>
+        public enum CrashReason
+        {
+            Unknown = 0,
+            UnhandledException = 1,
+            EnvironmentFailFast = 2,
+            InternalFailFast = 3,
+        }
+
+        /// <summary>
+        /// Type of runtime (currently only NativeAOT is used)
+        /// </summary>
+        public enum RuntimeType
+        {
+            Unknown = 0,
+            Desktop = 1,
+            NetCore = 2,
+            SingleFile = 3,
+            NativeAOT = 4,
+        }
+
+        /// <summary>
+        /// This block is passed to the Watson DOTNET CLRMA provider and must not have any GC references
+        /// </summary>
+        private struct TriageBlockBuffer
+        {
+            public const int BufferSize = 8192;
+            public fixed byte Buffer[BufferSize];
+        }
+
+        private bool _comma;
+        private int _currentBufferIndex;
+        private int _reservedBuffer;
+        private static TriageBlockBuffer _triageBuffer;
+
+        public CrashInfo()
+        {
+            _comma = false;
+            _currentBufferIndex = 0;
+            _reservedBuffer = 1;            // Reserve 1 byte for closing }
+        }
+
+        /// <summary>
+        /// Writes the opening bracket and header to triage buffer
+        /// </summary>
+        /// <param name="reason"></param>
+        /// <param name="message"></param>
+        public void Open(RhFailFastReason reason, string message)
+        {
+            // Write the opening bracket and basic header which should never fail
+            bool success = WriteChar('{');
+            Debug.Assert(success);
+            success = WriteHeader(reason, message);
+            Debug.Assert(success);
+        }
+
+        /// <summary>
+        /// Write the closing bracket. The triage buffer is ready.
+        /// </summary>
+        public void Close()
+        {
+            // Write the closing bracket which should never fail since it was reserved
+            _reservedBuffer = 0;
+            bool success = WriteChar('}');
+            Debug.Assert(success);
+        }
+
+        /// <summary>
+        /// Write the exception information with fallbacks if the info doesn't fit in the fixed size triage buffer.
+        /// </summary>
+        /// <param name="exception">exception object</param>
+        public void WriteExceptionInfo(Exception exception)
+        {
+            if (!WriteBlock("exception", '{', '}', () => WriteException(exception, int.MaxValue, 500, int.MaxValue)))
```

Comment:
These lambda delegates are going to bring in reflection stack and regress the size of basic hello world a lot. (Like going from 1.2MB to 1.9MB. This should stay away from delegates - it is also going to help with keeping the allocations in this code to minimum.